[TACHYON-108] Remote write support

gpang opened the pull request:
https://tachyon.atlassian.net/browse/TACHYON-108

This PR adds support for remote writes in Tachyon. The implementation is very simple, which means it NOT optimized at all. There are many opportunities to make everything faster and more efficient. Optimizations can be implemented in future PRs.

(LocalBlockOutStream.java was pretty much copied from BlockOutStream.java, but github doesn't show that)

calvinjia commented:
Jenkins, retest this please.

haoyuan commented:
Jenkins, test this please.

haoyuan commented:
Jenkins, test this please.

haoyuan commented:
Jenkins, test this please.

haoyuan commented:
Jenkins, test this please.

haoyuan commented:
Jenkins, test this please.

haoyuan commented:
Jenkins, test this please.

haoyuan commented:
Jenkins, test this please.

haoyuan commented:
Jenkins, test this please.

haoyuan commented on common/src/main/java/tachyon/network/protocol/RPCBlockWriteRequest.java:
```diff
@@ -0,0 +1,109 @@
+  private final DataBuffer mData;
+
+  public RPCBlockWriteRequest(long userId, long blockId, long offset, long length,
+                              DataBuffer data) {
```
format

gpang replied:
done.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java:
```diff
@@ -16,238 +16,62 @@
-    }
-
-    mWrittenBytes += len;
+  public BlockOutStream(TachyonFile tachyonFile, WriteType opType, TachyonConf tachyonConf) {
```
is `public` or `private`, given we have factory method?

gpang replied:
I made it protected, so subclasses can call it.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java:
```diff
@@ -0,0 +1,254 @@
+import tachyon.util.CommonUtils;
+
+/**
+ * <code>LocalBlockOutStream</code> implementation of TachyonFile. This class is not client facing.
```
NIT: I think the javadoc here is both confusing and probably wrong. Better improve the javadoc here.

gpang replied:
I updated the comments.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java:
```diff
@@ -0,0 +1,254 @@
+
+    if (!mTachyonFS.hasLocalWorker()) {
+      mCanWrite = false;
+      String msg = "The machine does not have any local worker.";
```
(1) why `msg` is needed?
(2) Assigning `mCanWrite` seems useless.

gpang replied:
I removed msg, and don't assign mCanWrite.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/RemoteBlockWriter.java:
```diff
@@ -0,0 +1,63 @@
+import tachyon.util.CommonUtils;
+
+/**
+ * The interface to read remote block from data server.
```
read -> write

gpang replied:
done.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java:
```diff
@@ -16,238 +16,62 @@
-import tachyon.util.CommonUtils;
 
 /**
  * <code>BlockOutStream</code> implementation of TachyonFile. This class is not client facing.
```
No more an implementation since this is an abstract class

gpang replied:
Updated comments.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java:
```diff
@@ -0,0 +1,254 @@
+      String msg = "The machine does not have any local worker.";
+      throw new IOException(msg);
+    }
+    mLocalFilePath = mTachyonFS.getLocalBlockTemporaryPath(mBlockId, initialBytes);
```
Better leave a TODO and state that LocalFileBlockWriter will be used instead of using java file API directly?

gpang replied:
Added the TODO.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java:
```diff
@@ -0,0 +1,213 @@
+/**
+ * <code>RemoteBlockOutStream</code> implementation of TachyonFile. This class is not client facing.
+ */
+public class RemoteBlockOutStream extends BlockOutStream {
```
`LocalBlockOutStream` is package-private (I thought that is by purpose, isn't it?) 
why this `RemoteBlockOutStream` is public?

apc999 commented on clients/unshaded/src/main/java/tachyon/client/TachyonFile.java:
```diff
@@ -497,7 +497,7 @@ boolean recache(int blockIndex) throws IOException {
           (int) mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES, Constants.MB) * 4;
       byte[] buffer = new byte[bufferBytes];
-      bos = new BlockOutStream(this, WriteType.TRY_CACHE, blockIndex, mTachyonConf);
+      bos = new LocalBlockOutStream(this, WriteType.TRY_CACHE, blockIndex, mTachyonConf);
```
should this be `LocalBlockOutStream` always? If `!mTachyonFS.hasLocalWorker()`, `new LocalBlockOutStream` will throw exceptions

gpang replied:
Yes, this can be either a Local one or Remote one. I changed it so it will choose accordingly. I had originally thought this was only possible locally, but remote should be possible. Thanks!

apc999 commented on common/src/main/java/tachyon/network/protocol/RPCMessage.java:
```diff
@@ -31,7 +31,10 @@
   // The possible types of RPC messages.
   public enum Type implements EncodedMessage {
-    RPC_BLOCK_REQUEST(0), RPC_BLOCK_RESPONSE(1);
+    RPC_BLOCK_REQUEST(0),
```
NIT: should this better be `RPC_BLOCK_READ_REQUEST` and `RPC_BLOCK_RESPONSE`? Maybe in another PR

gpang replied:
They can be renamed in a different PR. I left a few TODO's to rename them.

apc999 commented:
Put a bunch of comments. Thanks for the patch!

gpang commented:
Jenkins, test this please.

gpang commented:
Thanks @haoyuan and @apc999 for the reviews! I addressed your comments.

This PR can be continued to be reviewed, but it will be updated again after the big worker WIP is merged into master, to use the new APIs. This PR should only be merged in after that.

gpang commented:
This PR is now rebased onto master.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/TachyonFS.java:
```diff
@@ -642,23 +642,7 @@ public synchronized ClientFileInfo getFileStatus(int fileId, boolean useCachedMe
-      }
-    }
-
+    CommonUtils.createBlockPath(blockPath);
```
Perhaps this should be handled in the worker as well? Though does not have to be in this PR. Then createBlockPath can go into the (not yet merged) BlockWorkerUtils?

gpang replied:
Ok, I added a TODO for this.

calvinjia commented on servers/src/main/java/tachyon/worker/netty/DataServerHandler.java:
```diff
@@ -126,6 +137,48 @@ private void handleBlockRequest(final ChannelHandlerContext ctx, final RPCBlockR
+      ChannelFuture future = ctx.writeAndFlush(resp);
+      future.addListener(ChannelFutureListener.CLOSE);
+    }
+  }
```
Does the writer need to be closed?

gpang replied:
The writer is being closed now. Thanks!

gpang commented:
Thanks for the comments @calvinjia . I addressed them and added some more test cases.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java:
```diff
@@ -16,239 +16,63 @@
-    CommonUtils.putIntByteBuffer(mBuffer, b);
-    mWrittenBytes ++;
-  }
+  public abstract long getRemainingSpaceByte();
```
Please javadoc this method since it is public.

gpang replied:
done.

calvinjia replied:
Maybe its better to name it `getRemainingSpaceBytes`?

gpang replied:
Renamed.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java:
```diff
@@ -0,0 +1,254 @@
+    super(file, opType, tachyonConf);
+
+    if (!opType.isCache()) {
+      throw new IOException("LocalBlockOutStream only support WriteType.CACHE");
```
stating the writeType which triggers this exception in message?

gpang replied:
Added the opType in the message.

apc999 commented on clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java:
```diff
@@ -0,0 +1,213 @@
+  }
+
+  // Write data to the remote block.
+  private synchronized void writeToRemoteBlock(byte[] bytes, int offset, int length)
```
javadoc why we need `synchronized` here

gpang replied:
Added comment.

apc999 commented on integration-tests/src/test/java/tachyon/client/BlockOutStreamIntegrationTest.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Integration tests for <code>tachyon.client.BlockOutStream</code>.
+ */
+public class BlockOutStreamIntegrationTest {
```
If `LocalTachyonCluster` is unnecessary here (by creating `TachyonConf` directly), we may move this test into unittest.

gpang replied:
It turns out the local cluster is necessary, since in order to create a file, it needs to contact the master.

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockDataManager.java:
```diff
@@ -243,7 +243,22 @@ public String createBlock(long userId, long blockId, int tierAlias, long initial
       throws FileDoesNotExistException, IOException {
     BlockStoreLocation loc = BlockStoreLocation.anyDirInTier(tierAlias);
-    mBlockStore.createBlockMeta(userId, blockId, loc, initialBytes);
+    TempBlockMeta createdBlock = mBlockStore.createBlockMeta(userId, blockId, loc, initialBytes);
```
Currently it is a little bit confusing because both `createBlockRemote` and `getTempBlockWriterRemote` returns `BlockWriter`. I think we may change this method `createBlockRemote` to only create a block---without returning a `BlockWriter`. And we always use `getTempBlockWriterRemote` to return the `BlockWriter`

gpang replied:
Yes, this is better. I changed it.

apc999 commented on servers/src/main/java/tachyon/worker/netty/DataServerHandler.java:
```diff
@@ -126,6 +137,52 @@ private void handleBlockRequest(final ChannelHandlerContext ctx, final RPCBlockR
+        throw new IOException("Empty buffer to write.");
+      }
+
+      if (offset == 0) {
```
If we change the semantics of `createBlockRemote` by not returning `BlockWriter`, then we could probably simplify the code here a bit.

gpang replied:
It is only slightly less complicated, since space allocation is different in each case.

apc999 commented:
Thanks @gpang for the update. I left a bunch of comments---mostly readability issues.

gpang commented:
Thanks @apc999 for the comments! I addressed them and updated the PR.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java:
```diff
@@ -16,239 +16,66 @@
+  public static BlockOutStream get(TachyonFile tachyonFile, WriteType opType, int blockIndex,
+      long initialBytes, TachyonConf tachyonConf) throws IOException {
+
+    if (tachyonFile.mTachyonFS.hasLocalWorker()
```
Is `hasLocalWorker` a strong enough condition? Its possible the local worker cannot allocate space for the request?

gpang replied:
I think if the local worker cannot allocate space, it will fail later on. This condition is the similar condition as before (just threw an exception if there was no local worker).

calvinjia replied:
Oh I meant since we have remote write, the write can possibly still be served, but this is a feature that can be added later.

gpang replied:
Yeah, that could be a useful feature, but I don't think it can be done simply right now.

In the future, maybe we can make the client re-connect to a different worker if the local worker does not have space. This may have to touch higher levels of the code, like the TFS, and/or master.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java:
```diff
@@ -16,239 +16,66 @@
+    if (tachyonFile.mTachyonFS.hasLocalWorker()
+        && tachyonConf.getBoolean(Constants.USER_ENABLE_LOCAL_WRITE,
+            Constants.DEFAULT_USER_ENABLE_LOCAL_WRITE)) {
+      LOG.info("Writing with local stream.");
```
Could we add more detail to the two `LOG.info` statements (L65/69)?

gpang replied:
Added more info to the messages.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java:
```diff
@@ -0,0 +1,254 @@
+
+    if (!mTachyonFS.hasLocalWorker()) {
+      throw new IOException("The machine does not have any local worker.");
+    }
```
Is this check still necessary since we do everything through the `get` method?

gpang replied:
The check is not necessary. I added more comments stating that the get() should be used.

apc999 replied:
But it might be good to have a check here.
how about having `Preconditions.checkState(mTachyonFS.hasLocalWorker())` here?

gpang replied:
I added it in case it is constructed directly.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java:
```diff
@@ -0,0 +1,254 @@
+    CommonUtils.changeLocalFileToFullPermission(mLocalFilePath);
+    // use the sticky bit, only the client and the worker can write to the block
+    CommonUtils.setLocalFileStickyBit(mLocalFilePath);
+    LOG.info(mLocalFilePath + " was created!");
```
Is there any more info we could add to this log statement?

gpang replied:
Added some more details.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java:
```diff
@@ -0,0 +1,254 @@
+    LOG.info(mLocalFilePath + " was created!");
+    mAvailableBytes += initialBytes;
+
+    long allocateBytes = mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES, Constants.MB) + 4L;
```
Do we need the `+ 4L`? Also maybe make the `userFileBufferBytes` an instance variable since we use it multiple times (and the conf is presumably constant for now)?

gpang replied:
Most (if not all) of this file was just renamed from the original BlockOutStream, so these are artifacts of the existing file. I never knew why there was that extra 4, but I removed it.

I also added an instance variable for the buffer size.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java:
```diff
@@ -0,0 +1,254 @@
+
+  @Override
+  public void flush() throws IOException {
+    // Since this only writes to memory, this flush is not outside visible.
```
Does it make sense to just append the current buffer if it exists?

gpang replied:
Done.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java:
```diff
@@ -0,0 +1,224 @@
+  private final Closer mCloser;
+  private final RemoteBlockWriter mRemoteWriter;
+
+  // The local write buffer to accumulate client writes.
```
Minor style comment, perhaps use `/** description */`?

gpang replied:
Done.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java:
```diff
@@ -0,0 +1,224 @@
+  // Total number of bytes written to block, but not necessarily flushed to the remote server.
+  private long mWrittenBytes = 0;
+  // True if stream is open.
+  private boolean mOpen = false;
```
Can we change this to `mClosed` and reverse the logic to be consistent with the rest of the codebase?

gpang replied:
Done.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java:
```diff
@@ -0,0 +1,224 @@
+   * @param tachyonConf the TachyonConf instance for this file output stream.
+   * @throws IOException
+   */
+  RemoteBlockOutStream(TachyonFile file, WriteType opType, int blockIndex, long initialBytes,
```
Does `initialBytes` ever get used?

gpang replied:
No, so I added a comment about that.

Essentially, since the remote write api is really simple right now, the server side handles the allocation to match the size of the first write. If we ever wanted to add a richer api or add state to the data server, we could use this initialBytes parameter.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java:
```diff
@@ -0,0 +1,224 @@
+
+    // Create a local buffer.
+    mBufferBytes = mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES, Constants.MB);
+    long allocateBytes = mBufferBytes;
```
Why do we need this variable?

gpang replied:
We don't. I removed it.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java:
```diff
@@ -0,0 +1,224 @@
+    if (mOpen) {
+      mCloser.close();
+      mOpen = false;
+      if (mWrittenBytes > 0) {
```
What's the reasoning behind checking `mWrittenBytes` since this does not necessarily mean anything was written remotely?

gpang replied:
Thanks for catching this! I added a variable keeping track of the flushed bytes.

gpang commented:
I updated the PR addressing your comments, @calvinjia . Thanks!

calvinjia commented:
Jenkins, test this please.

gpang commented:
Jenkins, test this please.

apc999 commented:
+1 LGTM. Thanks for completing this important feature!

gpang commented:
Thanks for all the reviews!

calvinjia commented:
+1 LGTM

haoyuan commented:
Awesome!
